### PR TITLE
VIDEO: Add generic subtitles in SRT format

### PR DESCRIPTION
**Pull request by sev-**
There is an ongoing trend to add subtitles to video-based cutscenes in games.

This PR adds two classes `Video::SRTParser` and `Video::Subtitles` classes that do the following:

1. Load specified subtitles in SRT format
2. Plays them in a GUI overlay on top of the video in the specified bounding box
3. Colors are supported
4. TTF fonts are supported and they're drawn with outlines 

Demonstration on how to use it is in Groovie engine which was implemented in one of the re-releases, and I will add support for those versions eventually.

Comments are welcome.

### CODE REVIEW

**Comment by ccawley2011** [comment]
I don't think the overlay is a good fit for displaying subtitles, since backends aren't required to allow updating the game screen while the overlay is opened - in particular, the comments in common/system.h say this:

> For 'coolness' we usually want to have an overlay that is blended over the game graphics. On backends that support alpha blending, this is no issue but on other systems this needs some trickery.
>
> Essentially, we fake (alpha) blending on these systems by copying the current game graphics into the overlay buffer when activating the overlay, and then manually compose whatever graphics we want to show in the overlay.
> This works because we assume the game to be "paused" whenever an overlay is active.

**Comment by athrxx** [comment]
Displaying subtitles on a CLUT screen without the overlay is a pita, though. It would require to hack up the video player to reserve a text color.

Might be easier to just have a way to force updates of the normal screen.

Or we might just introduce another texture, only to be used for the subtitles. But that would also require backend upgrades...

File: video/subtitles.cpp [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+}
+
+void Subtitles::loadSRTFile(const char *fname) {
+	debug(1, "loadSRTFile('%s')", fname);
+
+	_loaded = _srtParser.parseFile(fname);
+}
+
+void Subtitles::setBBox(const Common::Rect bbox) {
```
Comment (mgerhardy) [inline video/subtitles.cpp:266]: could get passed by ref

File: video/subtitles.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+namespace Video {
+
+struct SRTEntry {
+	uint seq;
+	uint32 start;
+	uint32 end;
+
+	Common::String text;
+
+	SRTEntry(uint seq_, uint32 start_, uint32 end_, Common::String text_) {
```
Comment (mgerhardy) [inline video/subtitles.h:43]: text_ could get passed by ref

File: video/subtitles.cpp [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+	}
+
+	qsort(_entries.data(), _entries.size(), sizeof(SRTEntry *), &SRTEntryComparator);
+
+	debug(6, "SRTParser: Loaded %d entries", _entries.size());
+
+	return true;
+}
+
+Common::String SRTParser::getSubtitle(uint32 timestamp) {
```
Comment (mgerhardy) [inline video/subtitles.cpp:216]: method could be const

**Comment by sev-** [comment]
> I don't think the overlay is a good fit for displaying subtitles since backends aren't required to allow updating the game screen while the overlay is opened - in particular, the comments in common/system.h say this:

Well, in fact, I think we could safely drop that comment. Yes, when we added it 17 years ago, we were thinking about making it simpler for the backend developers. But since that, all the current backends do support displaying both overlays at the same time as they were following the SDL backend implementation.

We do use the GUI overlay, for instance, for the Cloud Sync icon these days.

So yes, it is safe to use it. To make sure it works as expected, I could say add another test in the graphics module.

The advantages of drawing subtitles on the overlay are the higher gfx resolution, often native and no headache with palettes.

**Comment by lephilousophe** [comment]
The N64 port seems to have such limitation: either the game is displayed, either the overlay.
It seems that game screen is copied to overlay when it is cleared.

**Comment by ccawley2011** [comment]
> The N64 port seems to have such limitation: either the game is displayed, either the overlay.
It seems that game screen is copied to overlay when it is cleared.

This is also the case for SurfaceSdlGraphicsManager as well. Because both of them exclusively use software rendering, changing them to use proper alpha transparency would likely hurt performance on platforms that would need software rendering.

> We do use the GUI overlay, for instance, for the Cloud Sync icon these days.

That's rendered using the OSD, not the overlay. This might be a better option here since it's typically implemented using small surfaces which should reduce the performance impact involved with alpha transparency. It also won't affect the mouse coordinate scale when it's active, so can be used for subtitles during gameplay as well as in cutscenes.

**Comment by sev-** [comment]
@BLooperZ already tried to implement subtitles for Feeble Files based on this.

The main finding is that indeed, even SDL backend does not support transparency, thus, a compromise was implemented: now there is a rectangle drawn around the subtitles, so on SDL backend it is neat and looking nice, while on backend with transparency it is still transparent. We would need to make it optional in the future.

I was looking into exposing Overlay similar to `lockScreen()`, but number of backends do not keep Overlay internal representation in a Surface, so that did not work well.

The same problem goes with OSD: it is not even implemented everywhere.

Thus, I am merging this as is, but we will need to add more flexibility to the rendering methods. I will do that in-tree eventually.

Thanks everyone for the feedback.

**Pull request merged**